### PR TITLE
Api search googletranslate consumption

### PR DESCRIPTION
**Pull request by talos**
This pull request is against `api-search-googletranslate`, as this is not yet merged into master.  It also includes a merge of the `definitions` branch, which will not need to be merged in separately (although definitions are not currently displayed).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,9 +64,10 @@ Then, in the root project directory:
 
     $ npm install mocha-phantomjs mocha
 
-Finally, in the root project directory:
+Finally, in the root project directory (while the test server is running):
 
-    $ node_modules/.bin/mocha-phantomjs -R dot http://localhost:8000/static/index.html --view 800x600 
+    $ node_modules/.bin/mocha-phantomjs -R dot http://localhost:8000/ --view 800x600 
+    $ node_modules/.bin/mocha-phantomjs -R dot http://localhost:8000/static/test/harness.html --view 800x600
```
Comment (antoviaque) [inline README.md:70]: Could be good to put into a `make test-client`, which would then allow to have a `make test` which runs both client and server tests.

Also, it would be good to be able to run the client tests without the server running - I think right now you need it because the original template is served by Django, right? (I'll have some comments about this below)

File: search/urls.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 urlpatterns = patterns('',
     url(r'^api/v1/search$', SearchAPIView.as_view(), name=u'api_search'),
-    url(r'^$', RedirectView.as_view(url=u'/static/index.html')),
+    url(r'^expression/', TemplateView.as_view(template_name=u'index.html')),
+    url(r'^$', TemplateView.as_view(template_name=u'index.html')),
```
Comment (antoviaque) [inline search/urls.py:15]: Why do we need to serve `index.html` through Django? You might have something in mind that I didn't think of, but I was hoping to be able to keep it fully static - that would dramatically speed up page load time without any need for caching if we can simply serve the client statically from a CDN.

I know this might not be possible for stuff like oauth, but I'd love to keep cases where we need to render templates on the server to a minimum - besides the load time concerns I was mentioning, this would also help us to ensure a clean separation between the client and the server logic.

File: static/js/routers/languagelearning.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        home: function () {
+            $header.hide();
+            this._searchBoxView.render().$el.detach().appendTo($horizon.show());
+        },
+
+        expression: function (expression) {
+            var self = this;
+
+            this.navigate('expression/' + expression);
```
Comment (antoviaque) [inline static/js/routers/languagelearning.js:37]: Does this work with all browsers, with unicode characters and white spaces? Honest question, since I have never tried this approach. I would have thought we would need to encode into utf8+urlencode, but maybe navigate() handles this already?

File: static/js/routers/languagelearning.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+            $horizon.hide();
+            this._searchBoxView.expression = expression;
+            this._searchBoxView.render().$el.detach().appendTo($header.show());
+
+            searchAdapter.search(expression).done(function (expressionModel) {
```
Comment (antoviaque) [inline static/js/routers/languagelearning.js:43]: It would be good to add an indicator that work is in progress while this request is being processed, and when it fails - I can especially see this here on my slow crappy connection, it's not obvious when/if the query completes (my connection drops regularly, or sometimes slows down to pings of several seconds).

Maybe changing the text color of `.main-container` to grey while it processes, and display an error message like "Could not connect - Retry" when the request does not complete? Would also likely be worth displaying the message from the server when the request completes but contains an error message.

File: README.md (hunk 1 above)
Comment (talos) [inline README.md:70]: Agreed -- and we shouldn't need Django running.

File: search/urls.py (hunk 2 above)
Comment (talos) [inline search/urls.py:15]: I only did this because I wasn't sure of a good way to do a reverse-proxy on `/index.html`.  I don't foresee any reason why we'd need to serve it through Django -- CDN would be best.  Is there a better way to path-hack this?

File: README.md (hunk 1 above)
Comment (talos) [inline README.md:70]: Extending the comment below, it would be best to test this against the "real" URL (`/` instead of `/static/index.html`).  However, I should be able switch this up to a fileserver and cut out Django, but we do run some risk of path problems.

File: static/js/routers/languagelearning.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        },
+
+        initialize: function (options) {
+            this._searchBoxView = new SearchBoxView({router: this});
+            this._expressionView = new ExpressionView({router: this});
+        },
+
+        home: function () {
+            $header.hide();
+            this._searchBoxView.render().$el.detach().appendTo($horizon.show());
```
Comment (antoviaque) [inline static/js/routers/languagelearning.js:31]: Backbone noob question again : ) What is `$el.detach()` for?

File: static/js/routers/languagelearning.js (hunk 3 above)
Comment (talos) [inline static/js/routers/languagelearning.js:37]: AFAIK JS is pretty good with this stuff -- the browser question is trickier, though.  Certainly unicode etc. characters are fine in Chrome, and should be in WebKit at large.  I'd be pretty surprised if modern IE or FF had problems, IE7 is a good question though.

File: static/js/routers/languagelearning.js (hunk 4 above)
Comment (talos) [inline static/js/routers/languagelearning.js:43]: Agreed!  I only didn't do this as it's a bit generic (we'll have other areas where we want an async 'loading' indicator), but should be on the list of things to do.

File: static/js/routers/languagelearning.js (hunk 5 above)
Comment (talos) [inline static/js/routers/languagelearning.js:31]: No problem!  This one is a little awkward -- what's going on here is that the view might have to migrate from the center of the page to the header, or vice-versa (here), so it's detached from its current portion of the DOM and re-attached to the correct one.  TBH I'm not a huge fan of doing this, I'll be thinking about other ways to handle it…

File: static/js/adapters/wiktionaryDefinitionProcessors.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+            $definitions = $('dd', $bedeutung.parent().next());
+            return _.map($definitions, function(def) {
+                return $(def).text();
+            });
+        },
+        'fr': function (page) {
+
+        }
+    };
+});
```
Comment (antoviaque) [inline static/js/adapters/wiktionaryDefinitionProcessors.js:25]: Not a big deal, but for clarity it would be better to put the definition files in another pull request imho - this might mean reverting those files in the current branch, and then recreating them in the definition branch, but otherwise it will be confusing when you'll create the pull request for the definitions, it will be missing some of the files.

File: search/urls.py (hunk 2 above)
Comment (antoviaque) [inline search/urls.py:15]: Ah, understood, I had assumed you also needed the templating capabilities. In this case it's ok to leave it as is yep. The best would probably be to write a dedicated view for this, to make make it more clear that the templating shouldn't be used, but for now that will do it just fine yep.

File: static/js/routers/languagelearning.js (hunk 5 above)
Comment (antoviaque) [inline static/js/routers/languagelearning.js:31]: Understood, it makes sense.

File: static/js/routers/languagelearning.js (hunk 3 above)
Comment (antoviaque) [inline static/js/routers/languagelearning.js:37]: Ok - actually that's a good question, which browsers do we want to support? IE7 has a very small market share now, not sure it's worth the pain (cf http://gs.statcounter.com/#browser_version_partially_combined-ww-monthly-201204-201304 ). IE8.0+?

File: static/test/adapters/search.test.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+   'mocha'
+], function (_, $, searchAdapter, ExpressionModel, expect) {
+
+    describe("Search API", function () {
+
+        it("provides a translation for an expression", function (done) {
+
+            var searchExpression = 'bom dia';
+
+            searchAdapter.search(searchExpression)
```
Comment (antoviaque) [inline static/test/adapters/search.test.js:18]: This still makes a real API call to the server, right? Would probably be good to use the same fake server approach as in `static/test/index.test.js` to keep it independent.

File: static/test/index.test.js [hunk 8]
```diff
@@ -50,5 +57,53 @@
...
+                                           "source": "pt",
+                                           "status": "success",
+                                           "target": "en"
+                                       })
+                                   ]);
+
+                $('input.search-text').val(expression);
+                $('form.search-form').trigger('submit');
+
+                server.respond();
```
Comment (antoviaque) [inline static/test/index.test.js:87]: Noob question again : ) How does this work? I can see that the server variable mocks the ajax call triggered line 85, but not sure how it it hooked up with Backbone? Does sinon overrides the jQuery ajax method?

File: static/js/routers/languagelearning.js (hunk 4 above)
Comment (antoviaque) [inline static/js/routers/languagelearning.js:43]: Ok to put the generic async loading indicator on the todo, but not sure how generic it will be - something generic to change the cursor might make sense, but for greying out the content being updated or displaying errors, it will depend on the context no?

**Comment by antoviaque** [comment]
Uh, it seems that the current pull request is marked as closed - I think github might have done this automatically when I deleted the `api-search-googletranslate` branch. I think github doesn't update the pull request with new commits when the pull request is closed, so you might have to reopen a pull request after all to integrate some of the changes above.

Btw did you wanted me to review the files related to the definitions too here? I didn't since I wasn't sure if you wanted to change it still, as it's not displayed on the results page yet - cf my remark above about creating a different pull request, but if you want I can also review it at the same time.

File: static/test/index.test.js [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*jslint browser:true*/
-/*globals describe, it, $, before, beforeEach, afterEach, expect*/
+/*globals describe, it, $, before, beforeEach, after afterEach, expect, sinon*/
```
Comment (antoviaque) [inline static/test/index.test.js:2]: Typo: forgot a ",", which makes jslint complain : )

File: README.md (hunk 1 above)
Comment (talos) [inline README.md:70]: The path problems are actually quite complex -- I think the best solution would be to have `make client-test` ensure some sort of reverse-proxy is running for the browser tests.  I messed around with using a SimpleHTTPServer on another port, but that can't do the reverse-proxying automatically.  So I think we should bunt on this for now, although I did create a `Makefile` so that this is no more than running `make test-client`.

File: static/js/routers/languagelearning.js (hunk 3 above)
Comment (talos) [inline static/js/routers/languagelearning.js:37]: I'd say we should support anything with at least 1% market share, which, thank heavens, no longer includes IE7.  So IE8+ sounds good to me (with an eye towards not even bothering with that once IE8 is below 1%, although that will optimistically probably be another two years.)

File: static/js/routers/languagelearning.js (hunk 4 above)
Comment (talos) [inline static/js/routers/languagelearning.js:43]: There's ways to make this reasonably generic in backbone -- i.e. to have the class your `View` descends from have a reasonably generic "I'm loading" feature, which grays out a relevant portion of the screen.

File: static/js/adapters/wiktionaryDefinitionProcessors.js (hunk 6 above)
Comment (talos) [inline static/js/adapters/wiktionaryDefinitionProcessors.js:25]: Hmm, the problem is that the definitions work was merged in here, and is pretty essential to some of the other stuff -- there won't need to be a separate pull request for definitions any more, and it should be clear from the commit history itself:

```
commit 45e5fde0c3ad3112529cc2b789a6c4976381cbfe
Merge: eab2603 25766c3
Author: John Krauss <irving.krauss@gmail.com>
Date:   Mon May 13 15:59:12 2013 -0400

    Merge branch 'definitions' into api-search-googletranslate-consumption

    Conflicts:
        README.md
```

exactly what happened.  Is that OK?

File: static/test/index.test.js (hunk 8 above)
Comment (talos) [inline static/test/index.test.js:87]: Sinon overrides `XMLHttpRequest` itself, along with everything descending from it -- jQuery's ajax method, Backbone's persistence (although we don't use it yet), etc.

File: static/test/index.test.js (hunk 9 above)
Comment (talos) [inline static/test/index.test.js:2]: Argle bargle. Fixed. :)

**Comment by talos** [comment]
That's annoying that GH closed this because the requested branch was deleted, but w/e.  I'll reopen against `master`.

Agreed it's a little messy with `definitions`, but I think we're okay because it's clear from the commit history itself (see above) how the `definitions` history plays out.  I still need to do other work, on `definitions-display`, for actually displaying the definitions we're getting.

File: static/js/routers/languagelearning.js (hunk 4 above)
Comment (antoviaque) [inline static/js/routers/languagelearning.js:43]: Oh, ok. Perfect then. Just remember to create a task for the loading to not forget : )

File: static/js/adapters/wiktionaryDefinitionProcessors.js (hunk 6 above)
Comment (antoviaque) [inline static/js/adapters/wiktionaryDefinitionProcessors.js:25]: Yup it's fine here, no need to jump through too much hoops for the definitions, especially since this is going to migrate to the server soon.

Just for the future, when possible try to keep one pull request = one feature, otherwise it makes it a bit harder to review. Typically here, if there was a dependency on the work from definitions for the current pull request, there could have been a first pull request for the definitions themselves, then another one for the googletranslate consumption. Would that make sense?

File: static/test/adapters/search.test.js (hunk 7 above)
Comment (antoviaque) [inline static/test/adapters/search.test.js:18]: This is still an issue, but since you mentioned you'd work on this during the meeting I'll still merge to master as is in the meantime, to not block this pull request.

File: static/test/index.test.js (hunk 8 above)
Comment (antoviaque) [inline static/test/index.test.js:87]: Understood - makes sense. Thanks for the explanation!
